File: csharp/AppEncryption/Crypto/NonceGenerator.cs
using System;
using System.Security.Cryptography;

namespace GoDaddy.Asherah.Crypto
{
  public class NonceGenerator
  {
    private const int BitsPerByte = 8;
    private readonly RandomNumberGenerator randomNumberGenerator = RandomNumberGenerator.Create();

    /// <summary>
    /// Generates a nonce.
    /// </summary>
    ///
    /// <param name="bits">Number of bits to use for nonce generation.</param>
    /// <returns>An array of nonce bytes.</returns>
    /// <exception cref="ArgumentException">If bits is not a multiple of 8.</exception>
    public byte[] CreateNonce(int bits)
    {
      if (bits % BitsPerByte != 0)
      {
        throw new ArgumentException($"Bits parameter must be a multiple of {BitsPerByte}");
      }

      byte[] keyBytes = new byte[bits / BitsPerByte];
      randomNumberGenerator.GetBytes(keyBytes);

      return keyBytes;
    }
  }
}

File: java/app-encryption/src/main/java/com/godaddy/asherah/crypto/NonceGenerator.java
package com.godaddy.asherah.crypto;

import java.security.SecureRandom;

public class NonceGenerator {
  // TODO Add ThreadLocal wrapper. Maybe consider adding periodic reseeding of this
  private final SecureRandom secureRandom = new SecureRandom();

  /**
   * Generates a nonce.
   * @param bits Number of bits to use for nonce generation.
   * @return An array of nonce bytes.
   */
  public byte[] createNonce(final int bits) {
    if (bits % Byte.SIZE != 0) {
      throw new IllegalArgumentException("Bits parameter must be multiple of 8");
    }

    byte[] keyBytes = new byte[bits / Byte.SIZE];
    secureRandom.nextBytes(keyBytes);
    return keyBytes;
  }
}
